src/node/redeem.rs

```diff
@@ -188,4 +189,51 @@ impl<J: Jet> RedeemNode<J> {
             |_, _| Ok(NoWitness),
         )
     }
+
+    /// Decode a Simplicity program from bits, including the witness data.
+    pub fn decode<I: Iterator<Item = u8>>(bits: &mut BitIter<I>) -> Result<Arc<Self>, Error> {
+        // 1. Decode program without witnesses as CommitNode
+        let commit = crate::decode::decode_program(bits)?;
+
+        // 2. Convert to RedeemNode, reading witnesses as we go
+        let witness_len = if bits.read_bit()? {
+            bits.read_natural(None)?
+        } else {
+            0
+        };
+        let witness_start = bits.n_total_read();
+
+        // Importantly, we do this directly using convert() rather than calling
+        // `.finalize()` because we need to use `InternalSharing` to make sure
+        // that we respect the sharing choices that were actually encoded in the
+        // bitstream.
+        let program: Arc<Self> = commit.convert::<InternalSharing, _, _, _, _>(
+            |data, converted| {
+                let converted_data = converted.map(|node| node.cached_data());
+                Ok(Arc::new(RedeemData::new(
+                    data.node.arrow().shallow_clone(),
+                    converted_data,
+                )))
+            },
+            |data, _| {
+                bits.read_value(&data.node.data.arrow().target)
+                    .map(Arc::new)
+            },
+        )?;
+
+        // 3. Check that we read exactly as much witness data as we expected
+        if bits.n_total_read() != witness_start + witness_len {
+            return Err(Error::InconsistentWitnessLength);
+        }
+
+        // 4. Check sharing
+        let mut imrs: HashSet<Imr> = HashSet::new();
+        for data in program.as_ref().post_order_iter::<InternalSharing>() {
+            if !imrs.insert(data.node.imr()) {
+                return Err(Error::SharingNotMaximal);
+            }
+        }
+
+        Ok(program)
```

Comment:
Using `is_shared_as` seems more canonical.
```
if program.as_ref().is_shared_as::<MaxSharing<Redeem, J>>() {
    Ok(program)
} else {
    Err(Error::SharingNotMaximal)
}
```

Reply:
We could use `is_shared_as` but that's at least twice as slow since it involves running two iterators at once. In practice it's probably more than twice as slow because it involves feeding IMRs into a hashmap which are 8 times as large as pointer IDs.

I'll add a comment explaining that they're equivalent though.

Reply:
Oh, maybe it's not twice as slow. It'll be slower but it's probably not much of a cost at all compared to the extra cost from using this iterator at all, versus an in-line algorithm.

At some point we should add some benchmarks and try to optimize this path, since it's used for decoding on-chain programs. But for now I'll leave it be.